### PR TITLE
[Lock] Split "StoreInterface" into multiple interfaces with less responsability

### PR DESCRIPTION
**Pull request by Simperfit**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | yes 
| BC breaks?    | no     
| Deprecations? | yes 
| Tests pass?   | yes    
| Fixed tickets | 
| License       | MIT
| Doc PR        | TODO 

We are removing the StoreInterface in order to split into multiple interface, it will help reduce de responsability of the StoreInterface.

Firstly, since not all stores needs to have all the methods of the StoreInterface, we can split this like this in order to limit the methods that are needed for each store.

Secondly, we add supportsX methods in order to avoid throwing exception when a store does not supports a feature it's easier an instance of the special interface or not, and it can return true/false on the support method.

### CODE REVIEW

File: src/Symfony/Component/Lock/ExpirableStoreInterface.php [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Lock;
+
+use Symfony\Component\Lock\Exception\LockConflictedException;
+use Symfony\Component\Lock\Exception\NotSupportedException;
+
+/**
+ * StoreInterface defines an interface to manipulate a lock store.
```
Comment (jderusse) [inline src/Symfony/Component/Lock/ExpirableStoreInterface.php]: `ExpirableStoreInterface`

File: src/Symfony/Component/Lock/BlockingStoreInterface.php [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Lock;
+
+use Symfony\Component\Lock\Exception\LockConflictedException;
+use Symfony\Component\Lock\Exception\NotSupportedException;
+
+/**
+ * StoreInterface defines an interface to manipulate a lock store.
```
Comment (jderusse) [inline src/Symfony/Component/Lock/BlockingStoreInterface.php]: `BlockingStoreInterface`

File: src/Symfony/Component/Lock/Lock.php [hunk 3]
```diff
@@ -36,13 +36,17 @@ final class Lock implements LockInterface, LoggerAwareInterface
...
+     * @param float|null                                                           $ttl         Maximum expected lock duration in seconds
+     * @param bool                                                                 $autoRelease Whether to automatically release the lock or not when the lock instance is destroyed
      */
-    public function __construct(Key $key, StoreInterface $store, float $ttl = null, bool $autoRelease = true)
+    public function __construct(Key $key, $store, float $ttl = null, bool $autoRelease = true)
     {
+        if ($store instanceof StoreInterface) {
+            @trigger_error(sprintf('"%s" is deprecated since Symfony 4.4 and has been splitted into "%s", "%s", "%s".', StoreInterface::class, PersistStoreInterface::class, ExpirableStoreInterface::class, BlockingStoreInterface::class), E_USER_DEPRECATED);
+        }
+
```
Comment (jderusse) [inline src/Symfony/Component/Lock/Lock.php:53]: You should assert here that the `$store` is an `PersistStoreInterface`.

File: src/Symfony/Component/Lock/Lock.php [hunk 4]
```diff
@@ -36,13 +36,17 @@ final class Lock implements LockInterface, LoggerAwareInterface
     private $dirty = false;
 
     /**
-     * @param Key            $key         Resource to lock
-     * @param StoreInterface $store       Store used to handle lock persistence
-     * @param float|null     $ttl         Maximum expected lock duration in seconds
-     * @param bool           $autoRelease Whether to automatically release the lock or not when the lock instance is destroyed
+     * @param Key                                                                  $key         Resource to lock
+     * @param PersistStoreInterface|BlockingStoreInterface|ExpirableStoreInterface $store       Store used to handle lock persistence
```
Comment (jderusse) [inline src/Symfony/Component/Lock/Lock.php]: IMHO it's not a "or" `|`, it should be a `PersistStoreInterface` + optional interfaces

File: src/Symfony/Component/Lock/Lock.php [hunk 5]
```diff
@@ -69,7 +73,7 @@ public function __destruct()
     public function acquire($blocking = false)
     {
         try {
-            if ($blocking) {
+            if ($blocking && $this->store instanceof BlockingStoreInterface) {
```
Comment (jderusse) [inline src/Symfony/Component/Lock/Lock.php]: `&& $this->store->supportsWaitAndSave()`

File: src/Symfony/Component/Lock/Lock.php [hunk 6]
```diff
@@ -121,7 +125,9 @@ public function refresh($ttl = null)
 
         try {
             $this->key->resetLifetime();
-            $this->store->putOffExpiration($this->key, $ttl);
+            if ($this->store instanceof ExpirableStoreInterface) {
```
Comment (jderusse) [inline src/Symfony/Component/Lock/Lock.php]: `&& $this->store->supportsPutOffExpiration`

File: src/Symfony/Component/Lock/Store/CombinedStore.php [hunk 7]
```diff
@@ -181,4 +184,20 @@ public function exists(Key $key)
...
+    {
+        return true;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function supportsWaitAndSave(): bool
+    {
+        return true;
```
Comment (jderusse) [inline src/Symfony/Component/Lock/Store/CombinedStore.php]: `false`

File: src/Symfony/Component/Lock/Store/CombinedStore.php [hunk 8]
```diff
@@ -181,4 +184,20 @@ public function exists(Key $key)
 
         return false;
     }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function supportsPutOffExpiration(): bool
+    {
+        return true;
```
Comment (jderusse) [inline src/Symfony/Component/Lock/Store/CombinedStore.php]: check if at least 1 store supports expiration.

At line  118 you should also check if the stores supports expiration and throws an exception if none of theme support it

File: src/Symfony/Component/Lock/Store/MemcachedStore.php [hunk 9]
```diff
@@ -21,7 +22,7 @@
  *
  * @author Jérémy Derussé <jeremy@derusse.com>
  */
-class MemcachedStore implements StoreInterface
+class MemcachedStore implements ExpirableStoreInterface
```
Comment (jderusse) [inline src/Symfony/Component/Lock/Store/MemcachedStore.php]: `PersistStoreInterface`

File: src/Symfony/Component/Lock/Store/RetryTillSaveStore.php [hunk 10]
```diff
@@ -99,4 +104,20 @@ public function exists(Key $key)
     {
         return $this->decorated->exists($key);
     }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function supportsPutOffExpiration(): bool
+    {
+        return true;
```
Comment (jderusse) [inline src/Symfony/Component/Lock/Store/RetryTillSaveStore.php]: `return $this->decorated instanceof ExpirableStoreInterface && $this->decorated->supportsPutOffExpiration();`

File: src/Symfony/Component/Lock/Store/RetryTillSaveStore.php [hunk 11]
```diff
@@ -99,4 +104,20 @@ public function exists(Key $key)
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function supportsWaitAndSave(): bool
+    {
+        return true;
```
Comment (jderusse) [inline src/Symfony/Component/Lock/Store/RetryTillSaveStore.php]: `$this->decorated instanceof BlockingStoreInterface && $this->decorated->supportsWaitAndSave()`

File: src/Symfony/Component/Lock/Store/RetryTillSaveStore.php [hunk 12]
```diff
@@ -81,7 +84,9 @@ public function waitAndSave(Key $key)
      */
     public function putOffExpiration(Key $key, $ttl)
     {
-        $this->decorated->putOffExpiration($key, $ttl);
+        if ($this->decorated instanceof ExpirableStoreInterface && $this->decorated->supportsPutOffExpiration()) {
```
Comment (jderusse) [inline src/Symfony/Component/Lock/Store/RetryTillSaveStore.php]: :+1:

Should do the same in `waitAndSave`

File: src/Symfony/Component/Lock/ExpirableStoreInterface.php [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+ *
+ * @author Hamza Amrouche <hamza.simperfit@gmail.com>
+ */
+interface ExpirableStoreInterface
```
Comment (jderusse) [inline src/Symfony/Component/Lock/ExpirableStoreInterface.php]: `Expiring`?

File: src/Symfony/Component/Lock/Store/SemaphoreStore.php [hunk 14]
```diff
@@ -98,10 +101,18 @@ public function delete(Key $key)
     }
 
     /**
-     * {@inheritdoc}
+     * Extends the ttl of a resource.
```
Comment (jderusse) [inline src/Symfony/Component/Lock/Store/SemaphoreStore.php:104]: Is this docblock needed? What's about other stores?

File: src/Symfony/Component/Lock/Store/CombinedStore.php [hunk 15]
```diff
@@ -26,7 +28,7 @@
  *
  * @author Jérémy Derussé <jeremy@derusse.com>
  */
-class CombinedStore implements StoreInterface, LoggerAwareInterface
```
Comment (jderusse) [inline src/Symfony/Component/Lock/Store/CombinedStore.php:29]: This is a BC break for code like:
```
class Service 
{
  public function __construct(StoreInterface $store)
  {
  }
}
```

Could be fixed in FrameworkExtension by decorating the store with a `LegacyStore` which trigger the deprecation

File: src/Symfony/Component/Lock/Store/RetryTillSaveStore.php [hunk 16]
```diff
@@ -37,7 +40,7 @@ class RetryTillSaveStore implements StoreInterface, LoggerAwareInterface
      * @param int            $retrySleep Duration in ms between 2 retry
      * @param int            $retryCount Maximum amount of retry
      */
-    public function __construct(StoreInterface $decorated, int $retrySleep = 100, int $retryCount = PHP_INT_MAX)
+    public function __construct(PersistStoreInterface $decorated, int $retrySleep = 100, int $retryCount = PHP_INT_MAX)
```
Comment (jderusse) [inline src/Symfony/Component/Lock/Store/RetryTillSaveStore.php]: BC Break. you should allow both of theme

File: src/Symfony/Component/Lock/Store/RetryTillSaveStore.php [hunk 17]
```diff
@@ -81,7 +84,9 @@ public function waitAndSave(Key $key)
      */
     public function putOffExpiration(Key $key, $ttl)
     {
-        $this->decorated->putOffExpiration($key, $ttl);
+        if ($this->decorated instanceof ExpirableStoreInterface && $this->decorated->supportsPutOffExpiration()) {
+            $this->decorated->putOffExpiration($key, $ttl);
+        }
```
Comment (jderusse) [inline src/Symfony/Component/Lock/Store/RetryTillSaveStore.php:101]: else throw an exception

**Review by jderusse — COMMENTED** [review]
First review :-)

File: src/Symfony/Component/Lock/Store/SemaphoreStore.php (hunk 14 above)
Comment (Simperfit) [reply src/Symfony/Component/Lock/Store/SemaphoreStore.php:104]: I need to add it on other stores

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 18]
```diff
@@ -1600,9 +1607,14 @@ private function registerLockConfiguration(array $config, ContainerBuilder $cont
...
             if (\count($storeDefinitions) > 1) {
+
                 $combinedDefinition = new ChildDefinition('lock.store.combined.abstract');
                 $combinedDefinition->replaceArgument(0, $storeDefinitions);
-                $container->setDefinition('lock.'.$resourceName.'.store', $combinedDefinition);
+
+                $legacyDefinition = new ChildDefinition('lock.store.legacy.abstract');
+                $legacyDefinition->replaceArgument(0, $combinedDefinition);
+
+                $container->setDefinition('lock.'.$resourceName.'.store', $legacyDefinition);
```
Comment (jderusse) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:1627]: `$container->setDefinition('lock.'.$resourceName.'.persist_store', $definition);`

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 19]
```diff
@@ -1625,9 +1637,13 @@ private function registerLockConfiguration(array $config, ContainerBuilder $cont
                 $container->setAlias('lock.factory', new Alias('lock.'.$resourceName.'.factory', false));
                 $container->setAlias('lock', new Alias('lock.'.$resourceName, false));
                 $container->setAlias(StoreInterface::class, new Alias('lock.store', false));
+                $container->setAlias(PersistStoreInterface::class, new Alias('lock.persist_store', false));
```
Comment (jderusse) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:1651]: `lock.persist_store` is not defined

should deprecate `lock.store` and `lock.store.*`

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/lock.xml [hunk 20]
```diff
@@ -24,6 +24,12 @@
             <argument type="service" id="lock.strategy.majority" /> <!-- Strategy -->
         </service>
 
+        <service id="lock.store.legacy.abstract" class="Symfony\Component\Lock\Store\LegacyStore" abstract="true">
```
Comment (jderusse) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/lock.xml:47]: not used?

File: src/Symfony/Component/Lock/Factory.php [hunk 21]
```diff
@@ -26,8 +26,11 @@ class Factory implements LoggerAwareInterface
 
     private $store;
 
-    public function __construct(StoreInterface $store)
+    public function __construct($store)
     {
+        if ($store instanceof StoreInterface) {
+            @trigger_error(sprintf('"%s" is deprecated since Symfony 4.4 and has been splitted into "%s", "%s", "%s".', StoreInterface::class, PersistStoreInterface::class, ExpiringStoreInterface::class, BlockingStoreInterface::class), E_USER_DEPRECATED);
```
Comment (jderusse) [inline src/Symfony/Component/Lock/Factory.php]: This error should have been laready triggered when user autoload the interface

File: src/Symfony/Component/Lock/Store/CombinedStore.php [hunk 22]
```diff
@@ -104,9 +118,10 @@ public function putOffExpiration(Key $key, $ttl)
     {
         $successCount = 0;
         $failureCount = 0;
+
         $storesCount = \count($this->stores);
         $expireAt = microtime(true) + $ttl;
-
+        $atLeastOneStoreSupportsExpiration = $this->supportsPutOffExpiration();
```
Comment (jderusse) [inline src/Symfony/Component/Lock/Store/CombinedStore.php]: I would have trigger exceptiojn her.
```
 if (false === $atLeastOneStoreSupportsExpiration) {
    throw new NotSupportedException();
 }
```

File: src/Symfony/Component/Lock/Store/LegacyStore.php [hunk 23]
```diff
@@ -0,0 +1,122 @@
...
+use Symfony\Component\Lock\Exception\LockConflictedException;
+use Symfony\Component\Lock\Exception\NotSupportedException;
+use Symfony\Component\Lock\ExpiringStoreInterface;
+use Symfony\Component\Lock\Key;
+use Symfony\Component\Lock\PersistStoreInterface;
+use Symfony\Component\Lock\StoreInterface;
+use Symfony\Component\Lock\Strategy\StrategyInterface;
+
+/**
+ * CombinedStore is a StoreInterface implementation able to manage and synchronize several StoreInterfaces.
```
Comment (jderusse) [inline src/Symfony/Component/Lock/Store/LegacyStore.php]: nop

File: src/Symfony/Component/Lock/Store/LegacyStore.php [hunk 24]
```diff
@@ -0,0 +1,122 @@
...
+ *
+ * @author Jérémy Derussé <jeremy@derusse.com>
```
Comment (jderusse) [inline src/Symfony/Component/Lock/Store/LegacyStore.php]: nop

File: src/Symfony/Component/Lock/Store/LegacyStore.php [hunk 25]
```diff
@@ -0,0 +1,122 @@
...
+    {
+        $this->decorated->exists($key);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function supportsPutOffExpiration(): bool
+    {
+        if ($this->decorated instanceof ExpiringStoreInterface && $this->decorated->supportsPutOffExpiration()) {
```
Comment (jderusse) [inline src/Symfony/Component/Lock/Store/LegacyStore.php]: `return $this->decorated instanceof ExpiringStoreInterface && $this->decorated->supportsPutOffExpiration();`

File: src/Symfony/Component/Lock/Store/RetryTillSaveStore.php [hunk 26]
```diff
@@ -37,8 +42,16 @@ class RetryTillSaveStore implements StoreInterface, LoggerAwareInterface
...
      * @param int            $retryCount Maximum amount of retry
      */
-    public function __construct(StoreInterface $decorated, int $retrySleep = 100, int $retryCount = PHP_INT_MAX)
+    public function __construct($decorated, int $retrySleep = 100, int $retryCount = PHP_INT_MAX)
     {
+        if ($decorated instanceof StoreInterface) {
+            @trigger_error(sprintf('"%s" is deprecated since Symfony 4.4 and has been splitted into "%s", "%s", "%s".', StoreInterface::class, PersistStoreInterface::class, ExpiringStoreInterface::class, BlockingStoreInterface::class), E_USER_DEPRECATED);
+        }
+
+        if ($decorated instanceof PersistStoreInterface || $decorated instanceof StoreInterface) {
```
Comment (jderusse) [inline src/Symfony/Component/Lock/Store/RetryTillSaveStore.php]: `if NOT`

File: src/Symfony/Component/Lock/Store/RetryTillSaveStore.php [hunk 27]
```diff
@@ -59,29 +72,37 @@ public function save(Key $key)
      */
     public function waitAndSave(Key $key)
     {
-        $retry = 0;
-        $sleepRandomness = (int) ($this->retrySleep / 10);
-        do {
-            try {
-                $this->decorated->save($key);
+        if ($this->decorated instanceof BlockingStoreInterface && $this->decorated->supportsWaitAndSave()) {
```
Comment (jderusse) [inline src/Symfony/Component/Lock/Store/RetryTillSaveStore.php]: use early return
```
if (!$this->decorated instanceof BlockingStoreInterface || !$this->decorated->supportsWaitAndSave()) {
  throw new NotSupportedException();
}
```
rest of code should net change

**Review by jderusse — COMMENTED** [review]
another pass

**Pull request closed without merging**